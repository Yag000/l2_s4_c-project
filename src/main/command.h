#ifndef COMMAND_H
#define COMMAND_H

#include <stdbool.h>

typedef struct command
{
    char *name;
    int args_number;
    char **args;

} command;

command *create_command(char *, int, char **);
void destroy_command(command *);

int execute_command(const command *);

bool handle_number_of_args(unsigned, unsigned);
int write_result_command(char *);
int write_result_lines_command(size_t, char **);

int pwd(const command *);
int rm(const command *cmd);

#endif