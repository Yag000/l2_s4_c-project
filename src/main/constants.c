#include "constants.h"

bool interactive = false;

const int EXIT_PROGRAM_SUCCESS = -2;  // Error code that indicates that the program executed successfully and that
                                      // it should be terminated
const int SUCCESS = 0;                // Error code that indicates that the program executed successfully
const int FATAL_ERROR = -1;           // Error code that indicates that he program should stop
const int INVALID_NAME = 1;           // Error code that indicates that the name of a node is invalid
const int INVALID_PATH = 2;           // Error code that indicates that the path of a node is invalid
const int INVALID_NUMBER_OF_ARGS = 3; // Error code that indicates that the number of arguments is invalid
const int INVALID_SELECTION = 4;      // Error code that indicates that the element selected by the user is invalid
