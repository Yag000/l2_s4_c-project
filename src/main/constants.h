#ifndef CONSTANTS_H
#define CONSTANTS_H

#include <stdio.h>
#include <stdbool.h>

extern FILE *out_stream;
extern bool verbose;

#endif