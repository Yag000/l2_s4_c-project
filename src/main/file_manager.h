#ifndef FILE_MANAGER_H
#define FILE_MANAGER_H

#include <stdio.h>

FILE *open_file(const char *path, const char *mode);
int close_file(FILE *file, const char *path);
int reset_file(const char *path);

#endif