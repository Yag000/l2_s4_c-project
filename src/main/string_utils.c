#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "string_utils.h"

void update_index_to_next_word(string_iterator *iterator);
unsigned get_number_of_char(size_t size, char **words);

string_iterator *create_string_iterator(char *string, char delimiter)
{
    string_iterator *iterator = malloc(sizeof(string_iterator));
    assert(iterator != NULL);
    iterator->string = string;
    iterator->index = 0;
    iterator->delimiter = delimiter;
    update_index_to_next_word(iterator);
    return iterator;
}

void destroy_string_iterator(string_iterator *iterator)
{
    free(iterator);
}

/**
 * Returns true if there is another word in the string.
 */
bool has_next_word(const string_iterator *iterator)
{
    return iterator->string[iterator->index] != '\0';
}

/**
 * Returns the next word in the string. If there are no more words
 * in the string, NULL is returned.
 */
char *next_word(string_iterator *iterator)
{
    if (!has_next_word(iterator))
    {
        return NULL;
    }

    // We iterate through the string until we find a delimiter, i order to skip the word
    int start = iterator->index;
    while (iterator->string[iterator->index] != '\0' && iterator->string[iterator->index] != iterator->delimiter)
    {
        iterator->index++;
    }
    int end = iterator->index;
    update_index_to_next_word(iterator);
    int length = end - start;

    char *result = malloc(sizeof(char) * (length + 1));
    assert(result != NULL);
    result = memmove(result, iterator->string + start, length);
    result[length] = '\0';

    return result;
}

/**
 * Updates the index to the next word in the string.
 */
void update_index_to_next_word(string_iterator *iterator)
{
    while (iterator->string[iterator->index] != '\0' && iterator->string[iterator->index] == iterator->delimiter)
    {
        iterator->index++;
    }
}

/**
 * Strips the new line character from a string.
 */
char *strip_newline(char *str)
{
    int i = 0;
    int j = 0;
    while (str[i] != '\0')
    {
        if (str[i] != '\n')
        {
            str[j++] = str[i];
        }
        i++;
    }
    str[j] = '\0';
    str = realloc(str, sizeof(char) * (j + 1));
    return str;
}

/**
 * Return the number of char in an array of array of char
 */
unsigned get_number_of_char(size_t size, char **words)
{
    if (words == NULL)
    {
        return 0;
    }

    unsigned n = 0;

    for (unsigned i = 0; i < size; ++i)
    {
        n += strlen(words[i]);
    }

    return n;
}

char *concat_two_words_with_delimiter(const char *word1, const char *word2, char delimiter)
{
    char *words_concat;

    unsigned len_word1 = strlen(word1);
    unsigned len_word2 = strlen(word2);

    if (len_word1 + len_word2 == 0)
    {
        words_concat = malloc(sizeof(char));
        assert(words_concat != NULL);

        words_concat[0] = '\0';
        return words_concat;
    }

    words_concat = malloc(sizeof(char) * (len_word1 + len_word2) + 2);
    assert(words_concat != NULL);

    memmove(words_concat, word1, len_word1);
    words_concat[len_word1] = delimiter;
    memmove(words_concat + len_word1 + 1, word2, len_word2);
    words_concat[len_word1 + len_word2 + 1] = '\0';

    return words_concat;
}

/**
 * Return the string of all words concatenated, and delimited by delimiter
 */
char *concat_words_with_delimiter(size_t size, char **words, char delimiter)
{
    char *words_concat;

    if (size <= 0 || words == NULL)
    {
        words_concat = malloc(sizeof(char));
        assert(words_concat != NULL);

        words_concat[0] = '\0';
        return words_concat;
    }

    words_concat = malloc(sizeof(char) * get_number_of_char(size, words) + size);

    assert(words_concat != NULL);

    unsigned acc = 0;
    unsigned len_word;

    for (unsigned i = 0; i < size - 1; ++i)
    {
        len_word = strlen(words[i]);
        memmove(words_concat + acc, words[i], len_word);

        acc += len_word;
        words_concat[acc] = delimiter;
        ++acc;
    }
    len_word = strlen(words[size - 1]);
    memmove(words_concat + acc, words[size - 1], len_word);
    acc += len_word;

    words_concat[acc] = '\0';

    return words_concat;
}