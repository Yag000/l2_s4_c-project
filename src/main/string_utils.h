#ifndef STRING_UTILS_H
#define STRING_UTILS_H

#include <stdbool.h>

/*
A string iterator that can be used to iterate over a string
with a given delimiter. It will return the next word in the
string until there are no more words left. A word is
defined as a string between two delimiters or between a delimiter
and the start/end of the string.
*/
typedef struct string_iterator
{
    char *string;
    int index;
    char delimiter;
} string_iterator;

string_iterator *create_string_iterator(char *, char);
void destroy_string_iterator(string_iterator *);
bool has_next_word(const string_iterator *);
char *next_word(string_iterator *);
int get_number_of_words_left(const string_iterator *);

void strip_newline(char *);

unsigned get_number_of_char(size_t, char **);
char *concat_two_words_with_delimiter(const char *, const char *, char);
char *concat_words_with_delimiter(size_t, char **, char);

bool is_alphanumeric(const char *);
char *get_alloc_pointer_of_string(const char *);

bool starts_with(const char *, const char *);

#endif
