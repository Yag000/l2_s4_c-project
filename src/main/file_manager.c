#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>

#include "file_manager.h"
#include "constants.h"

static void print_error_message(const char *, const char *);

/*
Opens the stream, send an error message if it fails
*/
FILE *open_file(const char *path, const char *mode)
{
    FILE *file = fopen(path, mode);
    if (file == NULL)
    {
        print_error_message("Problème ouverture fichier ", path);
    }
    return file;
}

/*
Closes the stream, send an error message if it fails
*/
int close_file(FILE *file, const char *path)
{
    int exit_code = fclose(file);

    if (exit_code != 0)
    {
        print_error_message("Problème fermeture fichier ", path);
        return FATAL_ERROR;
    }
    return SUCCESS;
}

/*
Prints the error message and the path of the file where it happened an error
*/
static void print_error_message(const char *error_message, const char *path)
{
    char *message = malloc(sizeof(char) * (strlen(error_message) + strlen(path) + 1));
    assert(message != NULL);
    message = strcat(message, error_message);
    message = strcat(message, path);
    perror(message);
    free(message);
}
