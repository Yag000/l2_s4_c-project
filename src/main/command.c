#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "command.h"
#include "constants.h"
#include "string_utils.h"
#include "tree_dir_core.h"

command *create_command(char *name, int args_number, char **args)
{
    command *cmd = malloc(sizeof(command));
    assert(cmd != NULL);
    cmd->name = name;
    cmd->args_number = args_number;
    cmd->args = args;
    return cmd;
}

/*
Frees the memory allocated for a command.
*/
void destroy_command(command *cmd)
{
    free(cmd->name);
    for (int i = 0; i < cmd->args_number; i++)
    {
        free(cmd->args[i]);
    }
    free(cmd->args);
    free(cmd);
}

/*
Prints a command if verbose mode is enabled.
*/
static void print_command(const command *cmd)
{
    if ((out_stream == stdout && interactive) || !verbose)
    {
        return;
    }

    print_command_header();

    fputs(cmd->name, out_stream);
    for (int i = 0; i < cmd->args_number; i++)
    {
        fputs(" ", out_stream);
        fputs(cmd->args[i], out_stream);
    }
    fputs("\n", out_stream);
}

/*
Prints the coimmand header. The command header is the current node path followed by a $.
*/
void print_command_header() { print_command_header_with_stream(out_stream); }

/*
Prints the coimmand header. The command header is the current node path followed by a $.
It will print the header in the given stream.
*/
void print_command_header_with_stream(FILE *stream)
{
    if (current_node != NULL)
    {
        char *path = get_absolute_path_of_node(current_node);
        fputs(path, stream);
        free(path);
    }
    fputs("$ ", stream);
}

int print_command_in_record_file(command *cmd)
{
    if (cmd == NULL || command_record_stream == NULL)
    {
        return EXIT_PROGRAM_SUCCESS;
    }
    fputs(cmd->name, command_record_stream);

    for (int i = 0; i < cmd->args_number; i++)
    {
        fputs(" ", command_record_stream);
        fputs(cmd->args[i], command_record_stream);
    }
    fputs("\n", command_record_stream);

    return SUCCESS;
}

/*
Creates a command from a string iterator.
*/
command *get_command_from_string_iterator(string_iterator *iterator)
{
    if (!has_next_word(iterator))
    {
        perror("Empty iterator problem");
        return NULL;
    }

    char *command_ = next_word(iterator);

    int args_number = get_number_of_words_left(iterator);

    char **args = malloc(sizeof(char *) * args_number);
    assert(args != NULL);

    int i = 0;
    while (has_next_word(iterator))
    {
        args[i] = next_word(iterator);
        i++;
    }

    return create_command(command_, args_number, args);
}

/*
Returns true if the command name matches the given name.
*/
static bool is_command(const command *command, const char *name) { return strcmp(command->name, name) == 0; }

/*
Executes a command and returns the exit code.
*/
int execute_command(const command *cmd)
{
    print_command(cmd);

    if (is_command(cmd, "ls"))
    {
        return ls(cmd);
    }
    if (is_command(cmd, "cd"))
    {
        return cd(cmd);
    }
    if (is_command(cmd, "pwd"))
    {
        return pwd(cmd);
    }
    if (is_command(cmd, "mkdir"))
    {
        return mkdir(cmd);
    }
    if (is_command(cmd, "touch"))
    {
        return touch(cmd);
    }
    if (is_command(cmd, "mv"))
    {
        return mv(cmd);
    }
    if (is_command(cmd, "rm"))
    {
        return rm(cmd);
    }
    if (is_command(cmd, "cp"))
    {
        return cp(cmd);
    }
    if (is_command(cmd, "print"))
    {
        return print(cmd);
    }
    if (is_command(cmd, "echo"))
    {
        return echo(cmd);
    }
    if (is_command(cmd, "exit"))
    {
        return EXIT_PROGRAM_SUCCESS;
    }
    int error_message_size = 100 + strlen(cmd->name);
    char *error_message = malloc(sizeof(char) * (error_message_size + 1));
    snprintf(error_message, error_message_size, "Command not found : %s", cmd->name);
    write_result_command(error_message);
    free(error_message);
    return EXIT_FAILURE;
}

/*
Returns true if the expected number of args is the same of actual
*/
bool handle_number_of_args(unsigned expected, unsigned actual)
{
    if (expected != actual)
    {
        char *error_message = malloc(sizeof(char) * 100);
        snprintf(error_message, 100, "An incorrect number of arguments was given: %u instead of %u expected.", actual,
                 expected);
        write_result_command(error_message);
        free(error_message);
        return false;
    }

    return true;
}

/*
Returns true if the actual number of args is between under_limit and upper_limit
*/
bool handle_number_of_args_with_delimitation(unsigned under_limit, unsigned upper_limit, unsigned actual)
{
    if (actual < under_limit || actual > upper_limit)
    {
        char *error_message = malloc(sizeof(char) * 100);
        snprintf(error_message, 100,
                 "An incorrect number of arguments was given: %u instead of a number between %u and %u expected.",
                 actual, under_limit, upper_limit);
        write_result_command(error_message);
        free(error_message);
        return false;
    }

    return true;
}

/*
Writes the string of result in the out_stream
*/
int write_result_command(char *result)
{
    // We need to also write the result in stdout if we are in interactive mode.
    // If not the user will not see the result.
    if (interactive && out_stream != stdout)
    {
        fputs(result, stdout);
        fputs("\n", stdout);
    }

    fputs(result, out_stream);
    fputs("\n", out_stream);

    return SUCCESS;
}

/*
Writes writes a series of strings in the out_stream
*/
int write_result_lines_command(size_t lines_number, char **results)
{
    for (unsigned i = 0; i < lines_number; i++)
    {
        write_result_command(results[i]);
    }
    return SUCCESS;
}
