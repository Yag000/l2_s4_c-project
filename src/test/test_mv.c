#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "../main/command.h"
#include "../main/constants.h"
#include "../main/file_manager.h"
#include "../main/string_utils.h"
#include "../main/tree_dir_core.h"
#include "test_core.h"

static void test_mv_function(test_info *);
static void test_command_mv_with_tree(test_info *);
static void test_error_of_mv(test_info *);

test_info *test_mv()
{
    // Create the test info
    print_test_header("mv");
    clock_t before = clock();
    test_info *info = create_test_info();

    // Add tests here
    test_mv_function(info);
    // End of tests
    info->time = clock_ticks_to_seconds(clock() - before);
    print_test_footer("mv", info);
    return info;
}

static void test_mv_function(test_info *info)
{
    test_command_mv_with_tree(info);
    test_error_of_mv(info);

    out_stream = stdout;
    out_stream_path = NULL;

    current_node = NULL;
}

static void test_command_mv_with_tree(test_info *info)
{
    current_node = create_basic_testing_tree();

    print_test_name("Testing mv with a tree");

    out_stream_path = "src/test/output/test_mv.txt";
    out_stream = open_file(out_stream_path, "w");

    char **tab_command = malloc(0);
    assert(tab_command != NULL);
    command *cmd_print = create_command(get_alloc_pointer_of_string("print"), 0, tab_command);

    tab_command = malloc(2 * (sizeof(char *)));
    assert(tab_command != NULL);
    command *cmd_mv = create_command(get_alloc_pointer_of_string("mv"), 2, tab_command);

    execute_command(cmd_print);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test11/test5";

    handle_boolean_test(true, execute_command(cmd_mv) == SUCCESS, __LINE__, __FILE__, info);
    execute_command(cmd_print);

    tab_command[0] = "/test";
    tab_command[1] = "/test11/test1";

    handle_boolean_test(true, execute_command(cmd_mv) == SUCCESS, __LINE__, __FILE__, info);
    execute_command(cmd_print);

    tab_command[0] = "/test11/test1/test9";
    tab_command[1] = "/test34";

    handle_boolean_test(true, execute_command(cmd_mv) == SUCCESS, __LINE__, __FILE__, info);
    execute_command(cmd_print);

    tab_command[0] = "/test11";
    tab_command[1] = "/NomChange11";

    handle_boolean_test(true, execute_command(cmd_mv) == SUCCESS, __LINE__, __FILE__, info);
    execute_command(cmd_print);

    tab_command[0] = get_alloc_pointer_of_string("/NomChange11");
    tab_command[1] = get_alloc_pointer_of_string("/test12/END");

    handle_boolean_test(true, execute_command(cmd_mv) == SUCCESS, __LINE__, __FILE__, info);
    execute_command(cmd_print);

    destroy_command(cmd_mv);
    destroy_command(cmd_print);

    close_file(out_stream, out_stream_path);

    destroy_tree();
}

static void test_error_of_mv(test_info *info)
{
    current_node = create_basic_testing_tree();

    print_test_name("Testing mv errors");

    out_stream_path = "src/test/output/test_mv_error.txt";
    out_stream = open_file(out_stream_path, "w");

    char **tab_command = malloc(2 * (sizeof(char *)));
    assert(tab_command != NULL);
    command *cmd = create_command(get_alloc_pointer_of_string("mv"), 2, tab_command);

    tab_command[0] = "test/test5/";
    tab_command[1] = "/test11/test5";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "test13";
    tab_command[1] = "/test11/test5";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "test/test5/test";
    tab_command[1] = "/test11/test5";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "test/test5";
    tab_command[1] = "/test11/test5/";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test11/test8/test11";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test11/test10/test11";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/testt/test5";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/testt/test5";

    handle_boolean_test(true, execute_command(cmd) == INVALID_PATH, __LINE__, __FILE__, info);

    tab_command[0] = "/test";
    tab_command[1] = "/test/test5/New";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    tab_command[0] = "/test11";
    tab_command[1] = "/test11/new";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test/test5/New";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    tab_command[0] = ".";
    tab_command[1] = "/test/test5/New";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    current_node = search_node_in_tree(current_node, "test/test5");

    tab_command[0] = "/test";
    tab_command[1] = "/test11/New";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test/test11/New";

    handle_boolean_test(true, execute_command(cmd) == INVALID_SELECTION, __LINE__, __FILE__, info);

    current_node = current_node->racine;

    tab_command[0] = "/test/test5";
    tab_command[1] = "/test/test5";

    handle_boolean_test(true, execute_command(cmd) == FATAL_ERROR, __LINE__, __FILE__, info);

    tab_command[0] = get_alloc_pointer_of_string("/test/test5");
    tab_command[1] = get_alloc_pointer_of_string("/test");

    handle_boolean_test(true, execute_command(cmd) == FATAL_ERROR, __LINE__, __FILE__, info);

    destroy_command(cmd);

    tab_command = malloc(0);
    assert(tab_command != NULL);
    cmd = create_command(get_alloc_pointer_of_string("mv"), 0, tab_command);
    handle_boolean_test(true, execute_command(cmd) == INVALID_NUMBER_OF_ARGS, __LINE__, __FILE__, info);

    destroy_command(cmd);

    tab_command = malloc(sizeof(char *) * 3);
    assert(tab_command != NULL);
    tab_command[0] = get_alloc_pointer_of_string("test");
    tab_command[1] = get_alloc_pointer_of_string("test");
    tab_command[2] = get_alloc_pointer_of_string("test");
    cmd = create_command(get_alloc_pointer_of_string("mv"), 3, tab_command);
    handle_boolean_test(true, execute_command(cmd) == INVALID_NUMBER_OF_ARGS, __LINE__, __FILE__, info);

    destroy_command(cmd);

    close_file(out_stream, out_stream_path);

    destroy_tree();
}