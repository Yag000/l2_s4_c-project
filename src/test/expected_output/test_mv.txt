/$ print
/ (D)
|-- test (D)
|   |-- test2 (F)
|   |-- test3 (D)
|   |-- test4 (F)
|   |-- test5 (D)
|   |   |-- test6 (F)
|   |   |-- test7 (D)
|   |   \-- test8 (F)
|   \-- test9 (F)
|-- test10 (F)
|-- test11 (D)
\-- test12 (D)
    |-- test13 (F)
    |-- test14 (F)
    \-- test15 (F)
/$ mv /test/test5 /test11/test5
/$ print
/ (D)
|-- test (D)
|   |-- test2 (F)
|   |-- test3 (D)
|   |-- test4 (F)
|   \-- test9 (F)
|-- test10 (F)
|-- test11 (D)
|   \-- test5 (D)
|       |-- test6 (F)
|       |-- test7 (D)
|       \-- test8 (F)
\-- test12 (D)
    |-- test13 (F)
    |-- test14 (F)
    \-- test15 (F)
/$ mv /test /test11/test1
/$ print
/ (D)
|-- test10 (F)
|-- test11 (D)
|   |-- test5 (D)
|   |   |-- test6 (F)
|   |   |-- test7 (D)
|   |   \-- test8 (F)
|   \-- test1 (D)
|       |-- test2 (F)
|       |-- test3 (D)
|       |-- test4 (F)
|       \-- test9 (F)
\-- test12 (D)
    |-- test13 (F)
    |-- test14 (F)
    \-- test15 (F)
/$ mv /test11/test1/test9 /test34
/$ print
/ (D)
|-- test10 (F)
|-- test11 (D)
|   |-- test5 (D)
|   |   |-- test6 (F)
|   |   |-- test7 (D)
|   |   \-- test8 (F)
|   \-- test1 (D)
|       |-- test2 (F)
|       |-- test3 (D)
|       \-- test4 (F)
|-- test12 (D)
|   |-- test13 (F)
|   |-- test14 (F)
|   \-- test15 (F)
\-- test34 (F)
/$ mv /test11 /NomChange11
/$ print
/ (D)
|-- test10 (F)
|-- test12 (D)
|   |-- test13 (F)
|   |-- test14 (F)
|   \-- test15 (F)
|-- test34 (F)
\-- NomChange11 (D)
    |-- test5 (D)
    |   |-- test6 (F)
    |   |-- test7 (D)
    |   \-- test8 (F)
    \-- test1 (D)
        |-- test2 (F)
        |-- test3 (D)
        \-- test4 (F)
/$ mv /NomChange11 /test12/END
/$ print
/ (D)
|-- test10 (F)
|-- test12 (D)
|   |-- test13 (F)
|   |-- test14 (F)
|   |-- test15 (F)
|   \-- END (D)
|       |-- test5 (D)
|       |   |-- test6 (F)
|       |   |-- test7 (D)
|       |   \-- test8 (F)
|       \-- test1 (D)
|           |-- test2 (F)
|           |-- test3 (D)
|           \-- test4 (F)
\-- test34 (F)
